graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,282 @@
+        });
+    }
+
+    /**
+     * Stores timing information and a map of nodes for field information.
+     */
+    private class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Node> nodeMap;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            // Create node map with the root node
+            nodeMap = new LinkedHashMap<>();
+            nodeMap.put(ExecutionPath.rootPath(), new Node(ExecutionPath.rootPath()));
+        }
+
+        Reports.Trace toProto() {
+            return Reports.Trace.newBuilder()
+                    .setStartTime(getStartTimestamp())
+                    .setEndTime(getNowTimestamp())
+                    .setDurationNs(getDuration())
+                    .setRoot(buildNodeTree())
+                    .build();
+        }
+
+        /**
+         * Adds node to nodeMap and recursively ensures that all parent node exist.
+         */
+        Node addNode(ExecutionPath path, Node node) {
+            if (!nodeMap.containsKey(path)) {
+                nodeMap.put(path, node);
+            }
+
+            List<Object> pathParts = path.toList();
+            ExecutionPath parentPath = ExecutionPath.fromList(pathParts.subList(0, pathParts.size() - 1));
+
+            if (!nodeMap.containsKey(parentPath)) {
+                addNode(parentPath, new Node(parentPath));
+            }
+
+            nodeMap.get(parentPath).addChild(node);
+            return node;
+        }
+
+        Node addNode(ExecutionStepInfo stepInfo, long startFieldNanos, long endFieldNanos) {
+            ExecutionPath path = stepInfo.getPath();
+            return addNode(path, new Node(
+                    stepInfo,
+                    startFieldNanos - startRequestNanos,
+                    endFieldNanos - startRequestNanos));
+        }
+
+        private Timestamp getStartTimestamp() {
+            return Timestamp.newBuilder()
+                    .setSeconds(startRequestTime.getEpochSecond())
+                    .setNanos(startRequestTime.getNano()).build();
+        }
+
+        private Timestamp getNowTimestamp() {
+            Instant now = Instant.now();
+            return Timestamp.newBuilder()
+                    .setSeconds(now.getEpochSecond())
+                    .setNanos(now.getNano())
+                    .build();
+        }
+
+        public long getDuration() {
+            return System.nanoTime() - startRequestNanos;
+        }
+
+        private Reports.Trace.Node buildNodeTree() {
+            return nodeMap.get(ExecutionPath.rootPath()).buildRecursive();
+        }
+    }
+
+    /**
+     * Wrapper for Node information with mutable lists for children and errors.
```

Comment:
Do we need to include this? I believe that the builder has accessors for adding children and errors.